File: products/catalyst/bootstrap/api/internal/handler/session_hint.go
// session_hint.go — the readable "a session exists" signal (#5940,
// UAT rows 3 + 91).
//
// ── The problem this exists to solve ─────────────────────────────────
//
// `catalyst_session` is HttpOnly by design and must stay that way. The
// marketplace (`core/marketplace/`) is a STATIC build on a SEPARATE host
// whose auth is deliberately client-side, so its JavaScript cannot read
// that cookie — `document.cookie` is empty even on the console's own
// origin. Measured on hw292 2026-08-10: an owner holding a live console
// session (`GET /api/v1/whoami` -> 200, tier `owner`, seconds later)
// opened their own voucher-redeem URL and was shown the "Sign up to
// redeem" stranger form; `redirectCount` was 0. The marketplace root
// behaved the same way — no returning-user redirect at all.
//
// Neither surface has a bug in its own logic. They have no readable
// signal, so the bounce cannot happen BY CONSTRUCTION.
//
// ── What this adds ───────────────────────────────────────────────────
//
// A second cookie, `catalyst_session_hint`, set alongside every
// `catalyst_session` and cleared alongside it. It is NOT HttpOnly, so a
// static page on a sibling host can read it, and it carries no secret:
//
//	v=1                (a session was minted by this Sovereign)
//	org=<slug>         (optional; the Organization the session is scoped
//	                    to — absent for a Sovereign-admin session)
//
// That is the whole value. No token, no JWT, no email, no jti, no
// expiry-bearing claim. It is a hint, never a credential: nothing on the
// server ever reads it back, so forging it grants a visitor exactly one
// thing — a redirect to a console that will then ask them to sign in.
// Authority stays entirely with the HttpOnly `catalyst_session`.
//
// ── Scope, stated honestly ───────────────────────────────────────────
//
// The hint is scoped to `sessionCookieDomain(r)` — EXACTLY the domain
// the real session cookie already spans. It therefore reaches every host
// the session itself reaches and not one host more:
//
//	console.<sov>            -> Domain=.<sov>        -> marketplace.<sov> READS IT
//	console.<slug>.<sov>     -> Domain=.<sov>        -> marketplace.<sov> READS IT
//	console.<slug>.<pool>    -> Domain=.<slug>.<pool>-> marketplace.<sov> does NOT
//
// The third row is a browser rule (a host cannot set a cookie for an
// unrelated registrable domain), not an omission. That persona signed up
// ON the marketplace origin and therefore holds the marketplace-origin
// `org-token`, which is the path that already works for it. The hint is
// strictly additive: it can only add a redirect where there was none.

package handler

import (
	"net/http"
	"net/url"
	"strings"
)

// SessionHintCookieName is the NON-HttpOnly companion to
// auth.SessionCookieName. Readable by client-side JS on any host inside
// the session cookie's domain, so a static sibling app (the marketplace)
// can tell that a session exists without ever seeing the session itself.
const SessionHintCookieName = "catalyst_session_hint"

// sessionHintVersion is stamped as `v=` so the reader can reject a value
// shape it does not understand instead of guessing at it.
const sessionHintVersion = "1"

// maxSessionHintOrgLen bounds the only variable-length field. A DNS label
// is 63 octets; an Org slug becomes one (`console.<slug>.<zone>`), so
// anything longer could not be a slug this Sovereign issued.
const maxSessionHintOrgLen = 63

// sanitizeHintOrg returns the Org slug if — and only if — it is a
// lowercase DNS label. Anything else yields "" and the hint is emitted
// without an `org` key.
//
// This is a whitelist rather than an escape: the value lands in a cookie
// that client-side JS splices into a hostname, so the set of acceptable
// characters is exactly the set legal in a DNS label. A rejected slug
// degrades to a session-exists-only hint, never to a malformed host.
func sanitizeHintOrg(org string) string {
	s := strings.ToLower(strings.TrimSpace(org))
	if s == "" || len(s) > maxSessionHintOrgLen {
		return ""
	}
	for i := 0; i < len(s); i++ {
		c := s[i]
		switch {
		case c >= 'a' && c <= 'z':
		case c >= '0' && c <= '9':
		case c == '-' && i != 0 && i != len(s)-1:
		default:
			return ""
		}
	}
	return s
}

// buildSessionHintValue renders the hint payload.
//
// The output is a sorted, URL-encoded key/value list — `v=1` alone for a
// Sovereign-admin session, `org=<slug>&v=1` for an Org-scoped one. Every
// key is enumerated here; there is no pass-through of caller data, so a
// new claim cannot leak into the hint by accident.
func buildSessionHintValue(org string) string {
	vals := url.Values{}
	vals.Set("v", sessionHintVersion)
	if slug := sanitizeHintOrg(org); slug != "" {
		vals.Set("org", slug)
	}
	return vals.Encode()
}

// setSessionHintCookie emits the hint next to a freshly minted session.
//
// Attributes mirror the session cookie (Path, Domain, Max-Age, Secure,
// SameSite) so the two live and die together in the browser's jar. The
// ONE deliberate difference is HttpOnly:false — that difference is the
// entire point, and it is safe only because the value is not a
// credential.
//
// Call this AFTER the session cookie's own http.SetCookie: HandlePinVerify
// and HandleOrgHandover both call w.Header().Del("Set-Cookie") immediately
// before minting the session (the TBD-F7/#1730 replay defence), which
// would otherwise drop the hint too.
func setSessionHintCookie(w http.ResponseWriter, org, domain string, maxAge int, secure bool) {
	http.SetCookie(w, &http.Cookie{
		Name:     SessionHintCookieName,
		Value:    buildSessionHintValue(org),
		Path:     "/",
		Domain:   domain,
		MaxAge:   maxAge,
		HttpOnly: false, // the point of this cookie — see the file header
		Secure:   secure,
		SameSite: http.SameSiteLaxMode,
	})
}

// buildClearSessionHintCookie returns the Set-Cookie header value that
// deletes the hint.
//
// Sign-out MUST clear this. A stale hint would bounce a signed-OUT
// visitor to a console that immediately asks them to sign in — an
// infinite door with no way back to the storefront, which is a worse
// customer outcome than the bug being fixed here.
//
// HttpOnly is deliberately absent — a clear-cookie must mirror the
// attributes the cookie was set with, and this one was set readable.
// SameSite stays Lax for the same reason.
//
// `maxAgeToken` is the literal Max-Age the caller's sibling cookies use on
// that response: the DELETE path contracts on `Max-Age=-1`, the SPA POST
// path on `Max-Age=0`. Both are non-positive and therefore identical to a
// browser (RFC 6265bis), so this is a wire-consistency choice — every
// Set-Cookie on one response reads the same way to anyone auditing it.
func buildClearSessionHintCookie(domain string, secure bool, maxAgeToken string) string {
	var b strings.Builder
	b.WriteString(SessionHintCookieName)
	b.WriteString("=; Path=/")
	if domain != "" {
		b.WriteString("; Domain=")
		b.WriteString(domain)
	}
	if secure {
		b.WriteString("; Secure")
	}
	b.WriteString("; SameSite=Lax; ")
	b.WriteString(maxAgeToken)
	return b.String()
}

File: core/marketplace/src/lib/sessionHint.ts
// sessionHint — the marketplace's readable "a session exists" signal
// (#5940, UAT rows 3 + 91).
//
// ── The fact this module exists to work around ───────────────────────
//
// `catalyst_session` is HttpOnly, and it must stay that way. The
// marketplace is a STATIC build on a SEPARATE host with client-side
// auth, so its JS cannot see that cookie: `document.cookie` reads EMPTY
// even on the console's own origin. Measured on hw292 2026-08-10 — an
// owner whose session was live (whoami 200, tier `owner`, seconds later)
// opened their own voucher-redeem link and got the "Sign up to redeem"
// stranger form, `redirectCount` 0. The marketplace root did nothing at
// all for the same visitor.
//
// catalyst-api now sets a second, NON-HttpOnly cookie next to every
// session (`products/catalyst/bootstrap/api/internal/handler/
// session_hint.go`). This module reads it.
//
// ── What a hint is, and what it is emphatically NOT ──────────────────
//
// It is a ROUTING signal, never a credential. Everything it can do is
// send the browser to a console — and that console then authenticates
// the visitor against the HttpOnly session exactly as it always did. A
// forged hint therefore buys an attacker one thing: a sign-in page.
//
// So this parser is deliberately paranoid in ONE direction only: it
// refuses to believe anything it cannot use safely. The `org` value is
// spliced into a hostname by the caller, so it is accepted only when it
// is a legal DNS label — a dot, a slash or a colon in that position
// could redirect a visitor off-Sovereign, which is the one outcome that
// would actually matter.

/** The cookie catalyst-api sets alongside every `catalyst_session`. */
export const SESSION_HINT_COOKIE = 'catalyst_session_hint';

/** The only payload version this reader understands. */
const SUPPORTED_VERSION = '1';

/**
 * A whole hint cannot be longer than this. The real value is ~15 bytes
 * (`org=<slug>&v=1`); anything approaching a token's length is something
 * this reader should refuse rather than try to interpret.
 */
const MAX_HINT_LENGTH = 128;

/** A legal lowercase DNS label — the only shape an Org slug can take. */
const DNS_LABEL = /^[a-z0-9]([a-z0-9-]{0,61}[a-z0-9])?$/;

export interface SessionHint {
  /**
   * The Organization the session is scoped to, or '' for a
   * Sovereign-admin session (which is scoped to the Sovereign itself and
   * has no Org slug to carry).
   */
  org: string;
}

/**
 * Parse a raw hint cookie VALUE.
 *
 * Returns null — meaning "treat this visitor as signed out" — when the
 * value is absent, over-long, of an unknown version, or shaped like a
 * credential. Returning null is always safe: the caller then renders the
 * public funnel, which is the behaviour that shipped before this fix.
 */
export function parseSessionHintValue(raw: string | null | undefined): SessionHint | null {
  if (!raw) return null;
  const value = raw.trim();
  if (!value || value.length > MAX_HINT_LENGTH) return null;

  // A dotted 3-segment value is JWT-shaped. The server never emits one,
  // so seeing one means either a bug or something being smuggled through
  // this channel — refuse either way rather than parse around it.
  if (value.split('.').length >= 3) return null;

  let params: URLSearchParams;
  try {
    params = new URLSearchParams(value);
  } catch {
    return null;
  }

  if (params.get('v') !== SUPPORTED_VERSION) return null;

  const org = (params.get('org') || '').trim().toLowerCase();
  // An unusable slug degrades to a session-exists-only hint — never to a
  // rejected hint, and never to a slug we splice into a host unchecked.
  return { org: DNS_LABEL.test(org) ? org : '' };
}

/**
 * Read the hint out of a `document.cookie` string.
 *
 * Cookie parsing is done by hand rather than with a regex over the whole
 * string because `catalyst_session_hint` and `catalyst_session` share a
 * prefix: a substring match would happily read the wrong cookie, and the
 * wrong cookie here is the HttpOnly one that will never be present.
 */
export function readSessionHint(cookieString: string | null | undefined): SessionHint | null {
  if (!cookieString) return null;
  for (const part of cookieString.split(';')) {
    const eq = part.indexOf('=');
    if (eq < 0) continue;
    if (part.slice(0, eq).trim() !== SESSION_HINT_COOKIE) continue;
    return parseSessionHintValue(decodeURIComponent(part.slice(eq + 1).trim()));
  }
  return null;
}

/** Browser-side convenience. Safe in SSR (returns null, no `document`). */
export function currentSessionHint(): SessionHint | null {
  if (typeof document === 'undefined') return null;
  try {
    return readSessionHint(document.cookie);
  } catch {
    return null;
  }
}
